subsys/retention/bootmode.c

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/retention/retention.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(bootmode, CONFIG_RETENTION_LOG_LEVEL);
+
+static const struct device *boot_mode_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_boot_mode));
+
+int bootmode_check(uint8_t boot_mode)
+{
+	int rc;
+
+	rc = retention_is_valid(boot_mode_dev);
+
+	if (rc == 1) {
+		uint8_t stored_mode;
+		int rc = retention_read(boot_mode_dev, 0, &stored_mode, sizeof(stored_mode));
```

Comment:
There's a check missing here for if `rc < 0`

Reply:
If rc is less then 0 then it will be returned directly to the caller, though I did miss removing the second rc definition

Reply:
ok maybe I'm missing something, bear with me
```
	int rc;

	rc = retention_is_valid(boot_mode_dev);

	if (rc == 1) {
		uint8_t stored_mode;

		rc = retention_read(boot_mode_dev, 0, &stored_mode, sizeof(stored_mode));

		if (stored_mode == boot_mode) {
			rc = 1;
		} else {
			rc = 0;
		}
	}
```
on `retention_read` if `rc < 0` I'm guessing `stored_mode != boot_mode` which will then set `rc = 0` so for the caller of `bootmode_check` the return is 0 but actually there was an issue with  `retention_read` that was not returned.

Do you agree ?

My suggested fix would be to just check for `rc > 0` then implement the logic ternary-equivalent to `rc = stored_mode == boot_mode ? 1 : 0;` (not saying you should use a ternary)

Reply:
Ah I see what you mean, you are correct, should be fixed now.